compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -1479,6 +1478,14 @@ impl<'a, 'tcx> InferCtxtPrivExt<'tcx> for InferCtxt<'a, 'tcx> {
                 // known, since we don't dispatch based on region
                 // relationships.
 
+                // Pick the first substitution that still contains inference variables as the one
+                // we're going to emit an error for. If there are none (see above), fall back to
+                // the substitution for `Self`.
+                let subst = {
+                    let substs = data.trait_ref.substs;
+                    substs.iter().find(|s| s.needs_infer()).unwrap_or_else(|| substs[0])
```

Comment:
We probably want to use something like `find(|s| s.needs_infer() && !matches!(s.unpack(), ty::GenericArgKind::Lifetime(_)))` as `emit_inference_failure_err` can't deal with lifetimes rn.

I expect this to ICE for a trait like `trait Foo<'a, T>` if both `'a` and `T` is unknown.What I am thinking of is something close to https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=66c6ee4894a18b7386ac056405b7cd02
```rust
trait Foo<'a, T> {
    fn test(&self) -> T {
        todo!()
    }
}

impl<'a> Foo<'static, u32> for &'a () {}

impl<'a> Foo<'a, i16> for &'a () {}

fn main() {
    let x = &();
    assert_eq!(x.test(), 7);
}
```

Reply:
I managed to get an ICE locally, and this actually picked something else up - it saw the Self subst `&()` as needing inference because of the lifetime on the reference, which we don't care about.

I'll switch `needs_infer` with `has_infer_types_or_conts`, which solves both problems since presumably no lifetime argument can satisfy that.

Reply:
yeah, that looks like the best solution to me :+1: